### PR TITLE
Issue 5279: Detect when the window surface it taken down on Android

### LINKED ISSUE #5100: Possible opengl reinit issue on Android
This issue focuses on the Android part. of this forum thread.

https://forum.defold.com/t/ios-crash-on-app-focus-gain-in-dmgraphics-clear-dmgraphics-context-unsigned-int-unsigned-char-unsigned-char-unsigned-char-unsigned-char-float-unsigned-int/64609/5?u=mathias_westerdahl

Currently, we have no clear repro.

### LINKED ISSUE #5279: Crash on Android (eglCreateWindowSurfaceTmpl)
**Describe the bug (REQUIRED)**
When user lock/unlock (miss lock by mistake) device pretty fast an app crashes.

**To Reproduce (REQUIRED)**
1. Use logcat/pidcat for logs, without logs crash hard to catch, because when you unlock apps restart and it seems like everything is fine
2. Run app on Android
3. Put lock button (hardware button) and unlock after a moment

It's not 100% repo case, repeat p.2-3 until crash

**Expected behavior (REQUIRED)**
App should work fine without crash

**Defold version (REQUIRED):**
  - 1.2.174 8f3e864464062e1b35c207521dc65dfd77899cdf

**Platforms (REQUIRED):**
 - Platforms: Android
 - OS: Android 10, 
 - Device: Samsung S10e (I can't reproduce on old Android device, maybe because it's not so responsive)

**Minimal repro case project (OPTIONAL):**
This is just empty project (Mobile template)
[TestMobileGame.zip](https://github.com/defold/defold/files/5397403/TestMobileGame.zip)

**Logs (OPTIONAL):**
```
           glfw-android  V  handleCommand: APP_CMD_RESUME
f67d601[DefoldActivity]  I  MSG_RESIZED_REPORT: frame=(0,0,1080,2280) ci=(0,117,0,144) vi=(0,0,0,144) or=1
           threaded_app  V  Pause: 0x7baa4dd6c0
                         V  activityState=13
           glfw-android  V  handleCommand: APP_CMD_PAUSE
     InputMethodManager  D  prepareNavigationBarInfo() DecorView@7c35685[DefoldActivity]
                         D  getNavigationBarColor() -855310
                         V  Starting input: tba=com.example.todo ic=null mNaviBarColor -855310 mIsGetNaviBarColorSuccess true , NavVisible : true , NavTrans : fals
                            e
                         D  startInputInner - Id : 0
                         I  startInputInner - mService.startInputOrWindowGainedFocus
f67d601[DefoldActivity]  I  Relayout returned: old=(0,0,1080,2280) new=(0,0,1080,2280) req=(1080,2280)8 dur=7 res=0x5 s={false 0} ch=true
                         I  ViewRootImpl >> surfaceDestroyed
           threaded_app  V  NativeWindowDestroyed: 0x7baa4dd6c0 -- 0x7b9ec53010
                         V  APP_CMD_TERM_WINDOW
           glfw-android  V  handleCommand: APP_CMD_TERM_WINDOW
                         V  destroy_gl_surface
           threaded_app  V  APP_CMD_TERM_WINDOW
           glfw-android  V  create_gl_surface
                 libEGL  E  eglCreateWindowSurfaceTmpl:890 error 300b (EGL_BAD_NATIVE_WINDOW)
           glfw-android  E  EGL_BAD_NATIVE_WINDOW
                   libc  F  ../lib/android/android_util.c:236: void create_gl_surface(_GLFWwin_android *): assertion "error == EGL_SUCCESS" failed
f67d601[DefoldActivity]  I  MSG_WINDOW_FOCUS_CHANGED 0 1
     InputMethodManager  D  prepareNavigationBarInfo() DecorView@7c35685[DefoldActivity]
                         D  getNavigationBarColor() -855310
           threaded_app  V  WindowFocusChanged: 0x7baa4dd6c0 -- 0
f67d601[DefoldActivity]  I  stopped(true) old=false
           threaded_app  V  Stop: 0x7baa4dd6c0
                 defold  I  INFO:CRASH: Successfully wrote Crashdump to file: /data/user/0/com.example.todo/files/_crash
                         E  ERROR:CRASH: CALL STACK:
                         E  # 0 pc        0x638 [vdso] <unknown>+0
                         E  # 1 pc      0x4235c ...oid.runtime/lib64/bionic/libc.so <unknown>+0
```

Full:
[crash-log.log](https://github.com/defold/defold/files/5397404/crash-log.log)

**Workaround (OPTIONAL):**
No

**Additional context (OPTIONAL):**

I noticed that when app crashes this message `glfw-android  V  handleCommand: APP_CMD_RESUME` appears earlier than `glfw-android  V  handleCommand: APP_CMD_LOST_FOCUS` (sometimes, when app crashes there is no `APP_CMD_LOST_FOCUS` at all)

### PR DESCRIPTION
**Pull request by JCash**

### CODE REVIEW

File: engine/glfw/lib/android/android_init.c [hunk 1]
```diff
@@ -203,6 +203,41 @@ GLFWAPI int32_t glfwAndroidWindowOpened()
     return _glfwWin.opened;
 }
 
+GLFWAPI int32_t glfwAndroidVerifySurface()
+{
+    // Although it's the wrong place to do a eglSwapbuffers, we're already handling a bad state from the last opengl error
+    // Verifying the state of the surface is worth it.
+    if (!eglSwapBuffers(_glfwWinAndroid.display, _glfwWinAndroid.surface))
```
Comment (JCash) [inline engine/glfw/lib/android/android_init.c:210]: This function is able to detect if the current window surface is ok.
We use it when checking the OpenGL errors, to decide if we should assert or not.

File: engine/glfw/lib/android/android_init.c [hunk 2]
```diff
@@ -230,7 +265,6 @@ void _glfwAndroidHandleCommand(struct android_app* app, int32_t cmd) {
             spinlock_lock(&_glfwWinAndroid.m_RenderLock);
 
             destroy_gl_surface(&_glfwWinAndroid);
-            _glfwWinAndroid.surface = EGL_NO_SURFACE;
```
Comment (JCash) [inline engine/glfw/lib/android/android_init.c:233]: Already reset in the `destroy_gl_surface()`

File: engine/glfw/lib/android/android_window.c [hunk 3]
```diff
@@ -343,8 +343,9 @@ void glfwAndroidFlushEvents()
                 create_gl_surface(&_glfwWinAndroid);
                 make_current(&_glfwWinAndroid);
                 update_width_height_info(&_glfwWin, &_glfwWinAndroid, 1);
+
+                computeIconifiedState();
```
Comment (JCash) [inline engine/glfw/lib/android/android_window.c:347]: No need to update unless anything changed.

File: engine/graphics/src/opengl/graphics_opengl.cpp [hunk 4]
```diff
@@ -188,18 +192,44 @@ static void LogGLError(GLint err, const char* fnname, int line)
...
-                err = glGetError(); \
+                LogGLError(err, __FUNCTION__, __LINE__); \
+                if (err == GL_OUT_OF_MEMORY) { \
+                    dmLogWarning("Signs of surface being destroyed. skipping assert.");\
+                    if (glfwAndroidVerifySurface()) { \
+                        assert(0); \
+                    } \
+                } else { \
+                    assert(0); \
+                } \
```
Comment (JCash) [inline engine/graphics/src/opengl/graphics_opengl.cpp:216]: This is the "magic" part.

File: engine/graphics/src/opengl/graphics_opengl.cpp [hunk 5]
```diff
@@ -188,18 +192,44 @@ static void LogGLError(GLint err, const char* fnname, int line)
...
             } \
         } \
-    }\
+    }
+
+#endif
+
+static void _ClearGLError()
```
Comment (JCash) [inline engine/graphics/src/opengl/graphics_opengl.cpp:223]: No need for this part to be a define.

**Review by britzl — APPROVED** [review]

**Pull request merged**